src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SingletonConnectionReader.cs

```diff
@@ -0,0 +1,612 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Runtime;
+using System.Runtime.CompilerServices;
+using System.Security.Authentication.ExtendedProtection;
+using System.ServiceModel;
+using System.ServiceModel.Channels.ConnectionHelpers;
+using System.ServiceModel.Description;
+using System.ServiceModel.Diagnostics;
+using System.ServiceModel.Dispatcher;
+using System.ServiceModel.Security;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+
+namespace System.ServiceModel.Channels
+{
+    internal abstract class SingletonConnectionReader
+    {
+        private IConnection _connection;
+        private bool _doneReceiving;
+        private bool _doneSending;
+        private bool _isAtEof;
+        private bool _isClosed;
+        private SecurityMessageProperty _security;
+        private object _thisLock = new object();
+        private int _offset;
+        private int _size;
+        private IConnectionOrientedTransportFactorySettings _transportSettings;
+        private Uri _via;
+        private Stream _inputStream;
+
+        protected SingletonConnectionReader(IConnection connection, int offset, int size, SecurityMessageProperty security,
+            IConnectionOrientedTransportFactorySettings transportSettings, Uri via)
+        {
+            _connection = connection;
```

Comment:
NIT: Would be good to have a Contract.Assert on connection != null (not necessary for security, we account for that and diverge code paths)